src/main/java/org/jitsi/videobridge/Endpoint.java

```diff
@@ -1300,24 +1300,50 @@ public void addChannel(ChannelShim channelShim)
      */
     public void removeChannel(ChannelShim channelShim)
     {
-        synchronized (channelShims)
+        if (channelShims.remove(channelShim))
         {
-            switch (channelShim.getMediaType())
+            if (channelShims.isEmpty())
+            {
+                expire();
+            }
+            else
             {
-                case AUDIO:
-                    acceptAudio = false;
-                    break;
-                case VIDEO:
-                    acceptVideo = false;
-                    break;
+                refreshMediaDirection();
             }
+        }
+    }
 
-            channelShims.remove(channelShim);
-            if (channelShims.isEmpty())
+    /**
+     * Refresh media direction based on direction of
+     * {@link ChannelShim}. An endpoint can accept
+     * media, if its {@link ChannelShim} has either
+     * 'sendrecv' or 'recvonly' media direction.
+     */
+    public void refreshMediaDirection()
+    {
+        boolean acceptAudio = false;
+        boolean acceptVideo = false;
+        for (ChannelShim channelShim : channelShims)
```

Comment:
what if we add another helper to `ChannelShim`: `supportsReceivedMedia` that does the `sendrecv`/`recvonly` check, and then do:
```
this.acceptAudio = channelShims.stream().filter(ChannelShim::supportsReceivingMedia).anyMatch(c -> c.getMediaType() == MediaType.AUDIO)
this.acceptVideo = channelShims.stream().filter(ChannelShim::supportsReceivingMedia).anyMatch(c -> c.getMediaType() == MediaType.VIDEO)
```

Reply:
@bbaldino I've added dedicated method `allowsReceivingMedia` into `ChannelShim`, I've chosen the different name because endpoint might support media, but temporary don't want receiving it. I've also prefer `foreach` loop over `Stream API`, because it allows to compute `accept*` fields in single pass over shims collection.

Reply:
yes, i think your naming is good for that reason.  as for saving the extra pass--i don't think it's a big deal as this should be an infrequent change and the list will be small so i think the more readable (in my opinion) stream version is nicer, but i don't feel strongly about it.